substrate/frame/conviction-voting/src/traits.rs

```diff
@@ -0,0 +1,90 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.use crate::AccountVote;
+
+//! Traits for conviction voting.
+
+use crate::AccountVote;
+use codec::{Decode, DecodeWithMemTracking, Encode, MaxEncodedLen};
+use frame_support::dispatch::DispatchResult;
+use scale_info::TypeInfo;
+use sp_runtime::RuntimeDebug;
+
+/// Represents the differents states of a referendum.
+/// None: The referendum is not started.
+/// Ongoing: The referendum is ongoing.
+/// Completed: The referendum is finished.
+#[derive(
+	Encode,
+	Decode,
+	DecodeWithMemTracking,
+	Copy,
+	Clone,
+	Eq,
+	PartialEq,
+	RuntimeDebug,
+	TypeInfo,
+	MaxEncodedLen,
+)]
+pub enum Status {
+	None,
+	Ongoing,
+	Completed,
+}
+pub trait VotingHooks<AccountId, Index, Balance> {
+	// Called when vote is executed.
+	fn on_vote(who: &AccountId, ref_index: Index, vote: AccountVote<Balance>) -> DispatchResult;
```

Comment:
```suggestion
	fn before_vote(who: &AccountId, ref_index: Index, vote: AccountVote<Balance>) -> DispatchResult;
```
The vote was not executed as it seems.